I pulled this version locally and I am trying to build my release notes. I am getting an error: 

```
No artifacter configured: every version tag will be considered a valid release
Git: Cloning <address> at /Users/rnowosielski/Documents/workspace/ivbcore/.
Git: Retrieving Git history...
Git: Retrieving versions (git tags)...
Make sure you first retrieve the promoted version: Promoted versions are empty, thus every tag matching will be considered as released

Generating using StashMarkdownWriter
Generating info for version 1.0.0
Traceback (most recent call last):
  File "/usr/local/bin/just_release", line 9, in <module>
    load_entry_point('JustReleaseNotes==0.1.8', 'console_scripts', 'just_release')()
  File "/Users/rnowosielski/Documents/workspace/JustReleaseNotes/JustReleaseNotes/command_line.py", line 26, in main
    generate_release_notes(args.config)
  File "/Users/rnowosielski/Documents/workspace/JustReleaseNotes/JustReleaseNotes/command_line.py", line 92, in generate_release_notes
    generateForOneWriter(generator, ticketProvider, writerConf["Provider"], os.path.dirname(path), os.path.basename(path))
  File "/Users/rnowosielski/Documents/workspace/JustReleaseNotes/JustReleaseNotes/command_line.py", line 44, in generateForOneWriter
    releaseNotes = generator.generateReleaseNotesByPromotedVersions(writer)
  File "/Users/rnowosielski/Documents/workspace/JustReleaseNotes/JustReleaseNotes/releaseNotes.py", line 56, in generateReleaseNotesByPromotedVersions
    if self.__repo.gitDatesByHash[hash] < self.__fromDate:
KeyError: u'e58ca29c5405f0a5fde04f63fab60218ad63f62d'
rake aborted!
Error while generating release notes with JustReleaseNotes
/Users/rnowosielski/Documents/workspace/ivbcore/Rakefile:152:in `block in <top (required)>'
Tasks: TOP => release_notes
(See full trace by running task with --trace)
rafalsnwmacbook:ivbcore rnowosielski$ 
```

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@ GitReleaseNotes/output/
 *.json
 JustReleaseNotes.egg-info/
 dist/
+build/

--- a/JustReleaseNotes/releaseNotes.py
+++ b/JustReleaseNotes/releaseNotes.py
@@ -1,34 +1,28 @@
+from datetime import datetime
+
+
 class ReleaseNotes:
 
     __PendingPromotionCaption = "Pending promotion"
     __conf = {}
-    __ticketsByVersion = {}
     __promotedVersionsInfo = {}
-    __ticketReplacements = {}
     __ticketProvider = {}
+    __fromDate = 0
 
     def __init__(self, conf, ticketProvider, repo, promotedVersionsInfo):
-        self.__ticketsByVersion = {}
         self.__conf = conf
         self.__repo = repo
         self.__promotedVersionsInfo = promotedVersionsInfo
-        self.__ticketProvider = ticketProvider;
+        self.__ticketProvider = ticketProvider
 
         self.__repo.checkout()
         self.__repo.retrieveHistory()
-        self.__repo.retrieveVersionsByGitHash( list(self.__promotedVersionsInfo.keys()) )
-        self.__computeTicketsByVersion()
-        
-    def __computeTicketsByVersion(self):
-        currentVersion = "latest"        
-        self.__ticketsByVersion[currentVersion] = []
-        for hash in self.__repo.gitCommitsList:
-            if hash in self.__repo.versionsByGitHash:
-                currentVersion = self.__repo.versionsByGitHash[hash]
-                self.__ticketsByVersion[currentVersion] = []
-            tickets = self.__ticketProvider.extractTicketsFromMessage(self.__repo.gitCommitMessagesByHash[hash])
-            self.__ticketsByVersion[currentVersion] = self.__ticketsByVersion[currentVersion] + tickets
-     
+        self.__repo.retrieveVersionsByGitHash(list(self.__promotedVersionsInfo.keys()))
+
+        if "OldestCommitToProcess" in conf["Source"]:
+            oldestCommitToProcess = conf["Source"]["OldestCommitToProcess"]
+            self.__fromDate = self.__repo.gitDatesByHash[oldestCommitToProcess]
+
     def __printVersionBlock(self, version, tickets, writer):
         date = "N/A"
         deps = {}
@@ -54,15 +48,17 @@ class ReleaseNotes:
         
         content = []
         for version in sortedVersions:
+
             if version in self.__promotedVersionsInfo or len(self.__promotedVersionsInfo.keys()) == 0:
                 print("Generating info for version " + version)
                 
             for hash in hashesInVersion[version]:
+                if self.__repo.gitDatesByHash[hash] < self.__fromDate:
+                    continue
                 if hash in hashAlreadySeen:
                     continue
                 hashAlreadySeen = hashAlreadySeen + [hash]
-                if hash not in self.__repo.gitCommitMessagesByHash:
-                    self.__repo.processCommit(hash)
+
                 ticketsSoFar += self.__ticketProvider.extractTicketsFromMessage(self.__repo.gitCommitMessagesByHash[hash])
 
             if version in self.__promotedVersionsInfo or len(self.__promotedVersionsInfo.keys()) == 0:

--- a/JustReleaseNotes/sourcers/GitRepo.py
+++ b/JustReleaseNotes/sourcers/GitRepo.py
@@ -25,6 +25,10 @@ class GitRepo:
         self.__directory = conf["Directory"]
         self.__repoX = ""
         self.__versionTagRegex = "^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+$"
+        self.__excludeCommitsWithMessageMatchingRegex = None
+
+        if "ExcludeCommitsWithMessageMatchingRegex" in conf:
+            self.__excludeCommitsWithMessageMatchingRegex = conf["ExcludeCommitsWithMessageMatchingRegex"]
 
         if "VersionTagRegex" in conf:
             self.__versionTagRegex = conf["VersionTagRegex"]
@@ -91,6 +95,11 @@ class GitRepo:
         self.__processCommit(self.__repoX.commit(commitHash))
 
     def __processCommit(self, commit):
+        if self.__excludeCommitsWithMessageMatchingRegex is not None:
+            p = re.compile(self.__excludeCommitsWithMessageMatchingRegex)
+            if p.match(commit.message):
+                return
+
         self.gitCommitMessagesByHash[commit.hexsha] = commit.summary + commit.message
         self.gitCommitsList.append(commit.hexsha)
         self.gitDatesByHash[commit.hexsha] = commit.authored_date

--- a/JustReleaseNotes/sourcers/GitRepo.py
+++ b/JustReleaseNotes/sourcers/GitRepo.py
@@ -74,9 +74,9 @@ class GitRepo:
         self.commitParents[commit.hexsha] = []
         for p in commit.parents:
             self.commitParents[commit.hexsha] = self.commitParents[commit.hexsha] + [p.hexsha]
-            self.setParents(p)
-            
-    def __getParentsListForVersion(self, hash, version, resultsSoFar):
+            self.__processCommit(p)
+
+    def __getParentsListForVersion(self, hash, resultsSoFar):
         if hash in resultsSoFar:
             return []
         if hash not in self.commitParents:
@@ -84,7 +84,7 @@ class GitRepo:
         results = [hash]
         for p in self.commitParents[hash]:
             if p not in self.versionsByGitHash:
-                results = results + self.__getParentsListForVersion(p, version, resultsSoFar + results)
+                results = results + self.__getParentsListForVersion(p, resultsSoFar + results)
         return results
 
     def processCommit(self, commitHash):
@@ -155,7 +155,7 @@ class GitRepo:
             else:
                 self.versionsByGitHash[hexsha] = version     
 
-            self.gitHistoryByVersion[version] = self.__getParentsListForVersion(hexsha,version, [])
+            self.gitHistoryByVersion[version] = self.__getParentsListForVersion(hexsha, [])
 
         self.__addHeadIfNotPresent()
         self.__optimizeHistoryByVersion()
@@ -165,4 +165,4 @@ class GitRepo:
         if hexsha not in self.versionsByGitHash:
             version = str(sys.maxsize)
             self.versionsByGitHash[hexsha] = version
-            self.gitHistoryByVersion[version] = self.__getParentsListForVersion(hexsha, version, [])
+            self.gitHistoryByVersion[version] = self.__getParentsListForVersion(hexsha, [])

--- a/README.rst
+++ b/README.rst
@@ -55,7 +55,7 @@ Configuration file is in flux. For now it is a json looking something like this:
                     "HtmlUrl" : ...,
                     "Authorization" : ...,
                     "Url" : ...,
-                    "WebImagesPath" : ...
+                    "WebImagesPath" : ...,
                     "TicketRegex" : ...
                 }],
                 "Releases" : {
@@ -65,14 +65,16 @@ Configuration file is in flux. For now it is a json looking something like this:
                     "StorageUrl" : ...
                 },
                 "Source" : {
-                    "Provider" : <source provider>
-                    "RepositoryUrl" : ...
-                    "Remote" : ...
-                    "Branch" : ...
-                    "VersionTagRegex" : "^([0-9]+\\.[0-9]+\\.[0-9]+)$"
+                    "Provider" : <source provider>,
+                    "RepositoryUrl" : ...,
+                    "Remote" : ...,
+                    "Branch" : ...,
+                    "VersionTagRegex" : "^([0-9]+\\.[0-9]+\\.[0-9]+)$",
+					"OldestCommitToProcess" : <commit hexsha>,
+					"ExcludeCommitsWithMessageMatchingRegex" : <regex>
                 },
                 "ReleaseNotesWriter" : [{
-                    "Provider" : ...,
+                    "Provider" : <notes writer provider>
                     "PathToSave" : ...
                 }]
             }
@@ -81,7 +83,7 @@ Configuration file is in flux. For now it is a json looking something like this:
 
 where
 
-``notes writer`` is either:
+``notes writer provider`` is either:
 
 - HtmlWriter
 - MarkdownWriter
@@ -96,7 +98,7 @@ where
 - JiraIssues
 - GitHubIssues
 
-It is also possible to specify an array of issue providers, then all of them will be used to retireve information about tickets.
+It is also possible to specify an array of issue providers, then all of them will be used to retrieve information about tickets.
 
 ``source provider`` is currently only:
 

--- a/tests/releaseNotes_Test.py
+++ b/tests/releaseNotes_Test.py
@@ -1,4 +1,5 @@
 import unittest
+from datetime import datetime
 import JustReleaseNotes
 from mock import Mock, MagicMock
 from JustReleaseNotes import releaseNotes
@@ -8,6 +9,8 @@ class ReleaseNote_Test(unittest.TestCase):
     def test_givenSimpleInputPrintVersionBlockGetsCalled(self):
 
         conf = {}
+        conf["Source"] = {}
+        conf["Source"]["OldestCommitToProcess"] = "ef334212ab2323a32323"
         promotedVersions = {}
         ticketProvider = Mock()
         ticketProvider.extractTicketsFromMessage = MagicMock(return_value=["TCKT-1"])
@@ -17,11 +20,16 @@ class ReleaseNote_Test(unittest.TestCase):
         writer = Mock()
         writer.printVersionBlock = MagicMock(return_value="empty")
         repo = Mock()
-        repo.gitCommitsList = ["ef334212ab2323a32323"]
-        repo.versionsByGitHash = { "ef334212ab2323a32323" : "1.0.1.2"}
-        repo.gitHistoryByVersion = {"1.0.1.2" : ["ef334212ab2323a32323"]}
-        repo.gitCommitMessagesByHash = { "ef334212ab2323a32323" : "Something about TCKT-1"}
-
+        repo.gitCommitsList = ["ef334212ab2323a32323",
+                               "as5d4a5sd4a5sd4a5sd4"]
+        repo.versionsByGitHash = {"ef334212ab2323a32323": "1.0.1.2",
+                                  "as5d4a5sd4a5sd4a5sd4": "1.0.1.1"}
+        repo.gitHistoryByVersion = {"1.0.1.2": ["ef334212ab2323a32323"],
+                                    "1.0.1.1": ["as5d4a5sd4a5sd4a5sd4"]}
+        repo.gitCommitMessagesByHash = {"ef334212ab2323a32323": "Something about TCKT-1",
+                                        "as5d4a5sd4a5sd4a5sd4": "Something about TCKT-0"}
+        repo.gitDatesByHash = {"ef334212ab2323a32323" : datetime.strptime("2015-12-12 12:12:12", "%Y-%m-%d %H:%M:%S").toordinal(),
+                               "as5d4a5sd4a5sd4a5sd4" : datetime.strptime("2015-11-11 11:11:11", "%Y-%m-%d %H:%M:%S").toordinal()}
 
         releaseNotes = JustReleaseNotes.releaseNotes.ReleaseNotes(conf, ticketProvider, repo, promotedVersions)
         releaseNotes.generateReleaseNotesByPromotedVersions(writer)

--- a/tests/sourcers/GitRepo_Test.py
+++ b/tests/sourcers/GitRepo_Test.py
@@ -4,11 +4,11 @@ from mock import Mock, MagicMock, mock_open, patch
 
 class Commit:
 
-    def __init__(self, hexsha):
+    def __init__(self, hexsha, message):
         self.hexsha = hexsha
         self.parents = []
         self.summary = ""
-        self.message = ""
+        self.message = message
         self.authored_date = ""
 
 class GitRepo_Test(unittest.TestCase):
@@ -30,9 +30,9 @@ class GitRepo_Test(unittest.TestCase):
     @patch("git.Repo.clone_from")
     @patch("git.Repo.iter_commits")
     def test_retrieveHistoryProcessesEachCommit(self, iter_commits, repo_mock):
-        commit1 = Commit("1")
-        commit2 = Commit("2")
-        commit3 = Commit("3")
+        commit1 = Commit("1", "Message 1")
+        commit2 = Commit("2", "Message 2")
+        commit3 = Commit("3", "Message 3")
         iter_commits.return_value = [commit1, commit3, commit2]
         mock = MagicMock()
         mock.iter_commits = iter_commits
@@ -51,10 +51,34 @@ class GitRepo_Test(unittest.TestCase):
         self.assertIn("3", repo.gitCommitsList)
         self.assertIn("3", repo.gitDatesByHash)
 
-    def test_setParentsTraersesChildrenAndDelasWithCycles(self):
+    @patch("git.Repo.clone_from")
+    @patch("git.Repo.iter_commits")
+    def test_retrieveHistoryIgnoresCommitsMatchConfiguredRegex(self, iter_commits, repo_mock):
+        commit1 = Commit("1", "Message 1")
+        commit2 = Commit("2", "Message 2")
+        commit3 = Commit("3", "Message 3")
+        iter_commits.return_value = [commit1, commit3, commit2]
+        mock = MagicMock()
+        mock.iter_commits = iter_commits
+        repo_mock.return_value = mock
+        conf = { "Directory" : "testDir", "RepositoryUrl" : "git://some.url/repo.git", "ExcludeCommitsWithMessageMatchingRegex" : "Message 2[.]*" }
+        repo = GitRepo.GitRepo(conf)
+        repo.checkout()
+        repo.retrieveHistory()
+        self.assertIn("1", repo.gitCommitMessagesByHash)
+        self.assertIn("1", repo.gitCommitsList)
+        self.assertIn("1", repo.gitDatesByHash)
+        self.assertNotIn("2", repo.gitCommitMessagesByHash)
+        self.assertNotIn("2", repo.gitCommitsList)
+        self.assertNotIn("2", repo.gitDatesByHash)
+        self.assertIn("3", repo.gitCommitMessagesByHash)
+        self.assertIn("3", repo.gitCommitsList)
+        self.assertIn("3", repo.gitDatesByHash)
+
+    def test_setParentsTraversesChildrenAndDealsWithCycles(self):
         conf = { "Directory" : "testDir", "RepositoryUrl" : "git://some.url/repo.git" }
         repo = GitRepo.GitRepo(conf)
-        commit = Commit("32423424af")
+        commit = Commit("32423424af", "Message")
         commit.message = "test message"
         commit.summary = "some summary"
         commit.authored_date = ""
@@ -64,13 +88,13 @@ class GitRepo_Test(unittest.TestCase):
         self.assertIn("32423424af", repo.gitCommitsList)
         self.assertIn("32423424af", repo.gitDatesByHash)
 
-    def test_setParentsTraersesChildrenAndDelasWithCycles(self):
+    def test_setParentsTraversesChildrenAndDealsWithCycles(self):
         conf = { "Directory" : "testDir", "RepositoryUrl" : "git://some.url/repo.git" }
         repo = GitRepo.GitRepo(conf)
-        commit = Commit("32423424af")
-        child = Commit("1")
+        commit = Commit("32423424af", "Message")
+        child = Commit("1", "Message")
         child.parents = [commit]
-        commit.parents = [ child, Commit("2") ]
+        commit.parents = [ child, Commit("2", "Message") ]
         repo.setParents(commit)
         self.assertEqual(2, len(repo.commitParents["32423424af"]))
         self.assertIn("32423424af", repo.commitParents)
@@ -79,7 +103,7 @@ class GitRepo_Test(unittest.TestCase):
     @patch("git.Repo.clone_from")
     @patch("os.path.isdir")
     @patch("os.makedirs")
-    def test_checkourt(self, makedirs_mock, isdir_mock, repo_mock):
+    def test_checkout(self, makedirs_mock, isdir_mock, repo_mock):
         isdir_mock.return_value = False
         conf = { "Directory" : "testDir", "RepositoryUrl" : "git://some.url/repo.git" }
         repo = GitRepo.GitRepo(conf)

